### PR TITLE
ENH: Add AVX2 and AVX512 functionality for numpy.sqrt

### PR DESCRIPTION
**Pull request by r-devulap**
In microbenchmarks which measured TSC cycles for SSE2, AVX2 and AVX512
versions of the sqrt function, AVX2 and AVX512 showed a 1.99x and 3.6x
performance improvement over SSE2. The experiments were performed on
Intel(R) Core(TM) i9-7900X CPU @ 3.30GHz. Detailed numbers are listed
below:

| Arraysize | Speed up with AVX2 | Speed up with AVX512 |
|-----------|--------------------|----------------------|
| 5000      | 1.99173            | 3.76                 |
| 20000     | 1.99784            | 3.69                 |
| 40000     | 1.99861            | 3.85                 |
| 80000     | 1.99891            | 3.85                 |
| 160000    | 1.99957            | 3.86                 |
| 320000    | 1.99521            | 3.86                 |
| 640000    | 1.99623            | 3.86                 |
| 1280000   | 1.99966            | 3.81                 |
| 2560000   | 1.96163            | 3.38                 |
| 5120000   | 1.92656            | 3.24                 |
| 10240000  | 1.92680            | 3.17                 |
| 20480000  | 1.92289            | 3.12                 |

### CODE REVIEW

File: numpy/core/src/umath/simd.inc.src [hunk 1]
```diff
@@ -941,6 +941,43 @@ sse2_binary_scalar2_@kind@_@TYPE@(npy_bool * op, @type@ * ip1, @type@ * ip2, npy
 static void
 sse2_sqrt_@TYPE@(@type@ * op, @type@ * ip, const npy_intp n)
 {
+#ifdef __AVX512F__
+    /* align output to VECTOR_SIZE_BYTES bytes */
+    LOOP_BLOCK_ALIGN_VAR(op, @type@, VECTOR_SIZE_BYTES) {
+        op[i] = @scalarf@(ip[i]);
+    }
+    assert(n < (VECTOR_SIZE_BYTES / sizeof(@type@)) || npy_is_aligned(&op[i], VECTOR_SIZE_BYTES));
```
Comment (charris) [inline numpy/core/src/umath/simd.inc.src]: Long lines here, break lines like in the other asserts.

**Comment by charris** [comment]
Couple of style nits, otherwise LGTM.

**Comment by juliantaylor** [comment]
I'd prefer this to use runtime selection, compile time only options are not very useful for the majority of users.

also interesting that avx finally does something for square root on newer cpus, it used to do nothing as it was just two sequential sse2 operations up to including haswell cpus.
I'll have a look at verifying these benchmarks on some newer cpus at work tomorrow.

File: numpy/core/src/umath/simd.inc.src (hunk 1 above)
Comment (r-devulap) [reply numpy/core/src/umath/simd.inc.src]: Fixed it with a new commit.

**Comment by juliantaylor** [comment]
hm I didn't have much time but a quick test on an `Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz` showed no difference in performance for sse2, avx2 and avx512 despite the code definitely running the right code paths.

**Comment by juliantaylor** [comment]
Also no difference in runtime or cycle count between sse and avx2 on a `AMD Ryzen 7 1700X`

**Comment by charris** [comment]
Perhaps these different experiences depend on the compiler/hardware. @juliantaylor I assume you are using (recent?) gcc. @r-devulap  Are you using the Intel compiler? What hardware?

**Comment by r-devulap** [comment]
I am using gcc 7.3.0 on Intel i9-7900X. I build NumPy using -march=native. I am basically measuring TSC cycles the function `DOUBLE_sqrt/FLOAT_sqrt` takes to execute (by using the `rdtsc `and `rdtscp `instructions before and after respectively). I did disable turbo and set intel_pstate governor to "performance" mode (to reduce variance across multiple runs). @juliantaylor How are you building NumPy?

**Comment by juliantaylor** [comment]
Also a gcc 7.3. The difference must be hardware, I checked the assembly that was produced and it looks fine. I also used rdtscp to check the cycle counts.
The `Intel(R) Xeon(R) Gold 6130` is a server cpu not a high end desktop variant but it is of the same generation and has the same amount of fma units. I kind of doubt that something as fundamental as a sqrt would differ between these types of cpus.
But I can be wrong. Though this type of major change must be documented somewhere.

**Comment by juliantaylor** [comment]
While I would like to understand why I cannot reproduce any speedups with avx2/avx512 and do not like the way this is implemented, it does not make things worse and I might just be using wrong hardware (though I have tested *alot* of different cpus over time ...).
So from my side this can be merged at the more active maintainers discretion.

(and again I want to plug my runtime variant of this in https://github.com/numpy/numpy/pull/11113 there are things to improve in that PR but base concept of runtime selected ufuncs works)

**Comment by charris** [comment]
I think I'll leave this open for a bit, it will be easy enough to backport. The main thing seems to be the runtime vs compile time variants, although it would be nice to track down the timing differences. @r-devulap Any chance you could try a lower end cpu in the same series?

**Comment by r-devulap** [comment]
My bad, @juliantaylor is right, I do not see any performance benefits either. I was using -O2 for compilation my micro-benchmarking and after debugging I realized gcc does not inline small functions consistently. If I inline them or use -O3 flag, then SSE2/AVX2/AVX512 all perform similarly. Vtune analysis shows that the function is memory bound and sqrt vectorization does not seem to help.

**Comment by charris** [comment]
I'll close this then. @r-devulap Thanks for the work and tracking down the cause of the timing discrepancy. @juliantaylor Thanks for the review.

**Pull request closed without merging**